File: 백준/이분탐색/1654 랜선자르기.py


# 최대 랜선의 길이를 탐색
# 가져온 랜선의 최대 길이를 구해서
# 그 사이에서 이분탐색

k, n = map(int, input().split(" "))
arr = []
for i in range(k):
    arr.append(int(input()))

def solution():
    # 나누기를 해서 몫들을 모두 합침.
    print(binary_search(n, 1, max(arr)))



def binary_search(target, start, end):
    mid = (start + end) // 2
    if start > end:
        return end

    if divide(mid) >= target:
        # 너무 작게 자른것, 더 크게 잘라야 함
        return binary_search(target, mid+1, end)
    else:
        return binary_search(target, start, mid-1)

def divide(value):
    sum = 0
    for i in arr:
        if i >= value:
            sum += i // value
    return sum


solution()

File: 백준/이분탐색/1654랜선자르기.js
const solution = (K, N, lans) => {
  let left = 1;
  let right = Math.max(...lans);
  let mid = 0;

  while (left <= right) {
    let count = 0;
    mid = parseInt((left + right) / 2, 10);
    for (let i = 0; i < lans.length; i++) {
      count += parseInt(lans[i] / mid, 10);
    }

    if (count >= N) {
      left = mid + 1;
    } else if (count < N) {
      right = mid - 1;
    }
  }
  console.log(right);
};

const fs = require('fs');

const stdin = (process.platform === 'linux'
  ? fs.readFileSync('/dev/stdin').toString()
  : `4 11
802
743
457
539
  `
).split('\n');

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const [K, N] = input().split(' ').map(Number);
const lans = [];
let flag = K;
while (flag--) {
  const lan = parseInt(input(), 10);
  lans.push(lan);
}
solution(K, N, lans);
